### PR TITLE
update RTMPose `SimCCPredictor`: expose `apply_softmax` and fix visibility thresholding

### PR DESCRIPTION
**Pull request by deruyter92**
**Motivation**
A regression analysis revealed that naive training results in reduced **mAP** in recent versions of DeepLabCut (only scores and mAP are affected; keypoint predictions remain the same, so mAR, RMSE are stable). 

**Cause:**
The cause could be pinpointed to the change in default value `apply_softmax=True` introduced in #3078. While this change is indeed required (output predictions should by default be normalized) this introduced two problems: 
- the combination of `apply_softmax=True` with `decode_beta=150` produces a sharp function which does not work well for all datasets. Since `apply_softmax` is not exposed in the default config yaml, this causes a silent change in performance.
- the visibility masking was not appropriately adjusted accordingly (`locs[vals <= 0.0] = -1` effectively is a no-op for the now normalized output values in range 0 to 1)

**Changes:**
- The `apply_softmax=True` is now exposed in the default config yaml, to facilitate reproducible benchmark results with frozen configs.  
- This PR fixes the visibility thresholding to a sensible value:  
```
threshold = 1.0 / simcc_x.shape[-1] if apply_softmax else 0.0
    locs[vals <= threshold] = -1
```

Note that the current PR does not alter the default value for the **decode_beta** parameter (might need further discussion and systematic testing), but for the current benchmarking results (see figure above), a value ~0.2 worked better than 300, and fully restored the original performance of before  #3078 

**Update**: a quick hyperparameter analysis was performed with different values of `decode_beta`. See figures at the bottom. **A separate PR will be opened to address this**

image

image

﻿
﻿

### CODE REVIEW

File: deeplabcut/pose_estimation_pytorch/models/predictors/sim_cc.py [hunk 1]
```diff
@@ -138,7 +138,8 @@ def get_simcc_maximum(
     mask = max_val_x > max_val_y
     max_val_x[mask] = max_val_y[mask]
     vals = max_val_x
-    locs[vals <= 0.0] = -1
+    threshold = 1.0 / simcc_x.shape[-1] if apply_softmax else 0.0
```
[the remark this answers is not in the document]
Comment (deruyter92) [reply deeplabcut/pose_estimation_pytorch/models/predictors/sim_cc.py:141]: hmm.. to me seems that this suggestion is just a different flavor of wrong. In practice it may not be necessary, but if needed I'd rather go for a complete separation of thresholds instead:

```python
# 1. Define independent thresholds
thresh_x = 1.0 / simcc_x.shape[-1] if apply_softmax else 0.0
thresh_y = 1.0 / simcc_y.shape[-1] if apply_softmax else 0.0

# 2. Identify where BOTH are valid
valid_mask = (max_val_x > thresh_x) & (max_val_y > thresh_y)

# 3. Proceed with merging
vals = torch.min(max_val_x, max_val_y)
locs[~valid_mask] = -1  # Invalidate if either axis is below its specific threshold
```

[the remark this answers is not in the document]
Comment (C-Achard) [reply deeplabcut/pose_estimation_pytorch/models/predictors/sim_cc.py:141]: `vals = torch.min(max_val_x, max_val_y)`
One small Q: Are these tensors or np arrays? (Conceptually irrelevant, I agree with your point)

**Review by C-Achard — APPROVED** [review]

**Review by AlexEMG — COMMENTED** [review]
Great!
